gematria/experiments/access_pattern_bm/contiguous_matrix_bm.cc

```diff
@@ -0,0 +1,85 @@
+#include <benchmark/benchmark.h>
+
+#include <iostream>
+
+#include "contiguous_matrix.h"
+
+namespace gematria {
+
+static void BM_FlushContiguousMatrixFromCache(benchmark::State &state) {
+  const std::size_t size = state.range(0);
+
+  // Create a random matrix
+  auto matrix = CreateRandomContiguousMatrix(size);
+
+  for (auto _ : state) {
+    FlushContiguousMatrixFromCache(matrix, size);
+  }
+
+  // Deallocate memory associated with the matrix
+  DeleteContiguousMatrix(matrix);
+}
+
+BENCHMARK(BM_FlushContiguousMatrixFromCache)->Range(1 << 4, 1 << 12);
+
+static void BM_ContiguousMatrix_NoFlush(benchmark::State &state) {
+  const std::size_t size = state.range(0);
+
+  // Create a random matrix
+  auto matrix = CreateRandomContiguousMatrix(size);
+  // auto mock = CreateRandomContiguousMatrix(size);
+
+  int sum = 0;
+  for (auto _ : state) {
+    // FlushContiguousMatrixFromCache(mock, size);
+
+    // Loop over the matrix, doing some dummy
+    // operations along the way.
+    for (int i = 0; i < size; ++i) {
+      for (int j = 0; j < size; ++j) {
+        sum += matrix[size * i + j];
+      }
+    }
+
+    benchmark::DoNotOptimize(sum);
+    sum = 0;
+  }
+
+  // Deallocate memory associated with the matrix
+  DeleteContiguousMatrix(matrix);
+  // DeleteContiguousMatrix(mock);
+}
+
+BENCHMARK(BM_ContiguousMatrix_NoFlush)->Range(1 << 4, 1 << 12);
+
+static void BM_ContiguousMatrix_Flush(benchmark::State &state) {
+  const std::size_t size = state.range(0);
+
+  // Create a random matrix
+  auto matrix = CreateRandomContiguousMatrix(size);
+
+  int sum = 0;
+  for (auto _ : state) {
+    FlushContiguousMatrixFromCache(matrix, size);
+
+    // Loop over the matrix, doing some dummy
+    // operations along the way.
+    for (int i = 0; i < size; ++i) {
+      for (int j = 0; j < size; ++j) {
+        sum += matrix[size * i + j];
+      }
+    }
```

Comment:
I ran both variants (this, and having both `i` and `j` go backwards) though a quick `perf stat` test, not expecting too much of a difference - but very interestingly, the forwards variant was measured to have 12,591,020 `mem_load_retired.l3_miss`es as compared to the backwards variant's 6,749,696 over the same 4s. Even accounting for the fact that each microbenchmark was not executed for the same number of iterations owing to how `benchmark` handles that, I wouldn't expect both measurements to vary by a factor that is  so strangely close to 2.

I think it would be really fun to understand what microarchitectural details result in that, and how.